### DISCUSSION TITLE
`string` to `Windows.Win32.Foundation.PCWSTR`

### DISCUSSION THREAD

**Discussion by myokeeh** [microsoft/CsWin32#912] [Q&A] [answered]
## Actual behavior

I currently have code that works simply by passing `string` arguments, but now it wants `Windows.Win32.Foundation.PCWSTR` arguments.

Worked fine on `0.2.188-beta`

## Expected behavior

I just want it working, so either need to have it accept `string` again or understand how to convert to `Windows.Win32.Foundation.PCWSTR`

### Context

- CsWin32 version: [e.g. `0.2.206-beta`]

**Comment by AArnott** [comment, marked as answer]
You can create a `PCWSTR` based on a `string` like this:

```cs
unsafe
{
    string message = "Hello World";
    fixed (char* pMessage = message)
    {
        PCWSTR pcwstr = new PCWSTR(pMessage);
        // use PCWSTR within the fixed block.
    }
}
```

**Reply by myokeeh** [reply]
Thanks. Looks simple enough. I'll try that.

**Reply by domenic** [reply]
Just found this... I tried to convert my project from "normal" P/Invoke to CSWin32 but now I have a bunch of this kind of code scattered through my codebase. It feels like a regression and I'm planning to go back to normal P/Invoke.

Has the situation gotten any better in the last 3 years?

**Before (normal P/Invoke)**

```cs
var wc = new WNDCLASSEXW
{
    cbSize = (uint)Marshal.SizeOf<WNDCLASSEXW>(),
    lpfnWndProc = Marshal.GetFunctionPointerForDelegate(s_wndProc),
    hInstance = GetModuleHandleW(null),
    hCursor = LoadCursorW(0, 32512), // IDC_ARROW
    hbrBackground = GetStockObject(4), // BLACK_BRUSH
    lpszClassName = WindowClassName
};

if (RegisterClassExW(ref wc) == 0)
{
    throw new Win32Exception(Marshal.GetLastPInvokeError());
}
```

**After (CsWin32)**

```cs
unsafe {
  fixed (char* classNamePtr = WindowClassName)
  {
      var wc = new WNDCLASSEXW
      {
          cbSize = (uint)Marshal.SizeOf<WNDCLASSEXW>(),
          lpfnWndProc = s_wndProc,
          hInstance = PInvoke.GetModuleHandle((PCWSTR)null),
          hCursor = PInvoke.LoadCursor(HINSTANCE.Null, (PCWSTR)((char*)32512)),  // IDC_ARROW
          hbrBackground = new HBRUSH(PInvoke.GetStockObject(GET_STOCK_OBJECT_FLAGS.BLACK_BRUSH).Value),
          lpszClassName = classNamePtr
      };

      if (PInvoke.RegisterClassEx(in wc) == 0)
      {
          throw new Win32Exception(Marshal.GetLastPInvokeError());
      }
  }
}
```
